devel/fortio/httprunner.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fortio
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"runtime"
+	"runtime/pprof"
+	"sort"
+)
+
+// Most of the code in this file is the library-fication of code originally
+// in cmd/fortio/main.go
+
+// HTTPRunnerResults is the aggregated result of an HTTPRunner.
+// Also is the internal type used per thread/goroutine.
+type HTTPRunnerResults struct {
+	client            Fetcher
+	RetCodes          map[int]int64
+	Sizes             *Histogram
+	HeaderSizes       *Histogram
+	DurationHistogram *Histogram
+	ActualQPS         float64
+}
+
+// Used globally / in TestHttp() TODO: change periodic.go to carry caller defined context
+var (
+	state []HTTPRunnerResults
+)
+
+// TestHTTP http request fetching. Main call being run at the target QPS.
+// To be set as the Function in RunnerOptions.
+func TestHTTP(t int) {
+	Debugf("Calling in %d", t)
+	code, body, headerSize := state[t].client.Fetch()
+	size := len(body)
+	Debugf("Got in %3d hsz %d sz %d", code, headerSize, size)
+	state[t].RetCodes[code]++
+	state[t].Sizes.Record(float64(size))
+	state[t].HeaderSizes.Record(float64(headerSize))
+}
+
+// HTTPRunnerOptions includes the base RunnerOptions plus http specific
+// options.
+type HTTPRunnerOptions struct {
+	RunnerOptions
+	URL         string
+	Compression bool   // defaults to no compression, only used by std client
+	StdClient   bool   // defaults to fast client
+	HTTP10      bool   // defaults to http1.1
```

Comment:
I might be tempted to use an enum here (in the future, you may want to add an http/2 client).